brainframe_qt/api_utils/connection_manager.py

```diff
@@ -0,0 +1,154 @@
+from dataclasses import dataclass
+from enum import Enum, auto
+from typing import Optional, Tuple
+
+from PyQt5.QtCore import pyqtSignal, QObject, QThread
+
+from brainframe.api import bf_errors, bf_codecs
+
+from brainframe_qt.api_utils import api
+from brainframe_qt.ui.resources.config import ServerSettings
+from brainframe_qt.util.secret import decrypt
+
+
+class ConnectionManager(QThread):
+    _CONNECTED_SLEEP_TIME = 100  # ms
+    _CONNECT_TIMEOUT = 5  # s
+
+    connection_state_changed = pyqtSignal(object)
+    license_state_changed = pyqtSignal(bf_codecs.LicenseInfo.State)
+
+    connection_error = pyqtSignal(Exception)
+
+    class ConnectionState(Enum):
+        UNCONFIGURED = auto()
+        """Credentials not accessed, or in need of update"""
+        UNCONNECTED = auto()
+        """No communication with server yet"""
+        LICENSE_UNVALIDATED = auto()
+        """License not yet validated"""
+        LICENSE_EXPIRED = auto()
+        """Current license is expired"""
+        LICENSE_INVALID = auto()
+        """Current license invalid"""
+        LICENSE_MISSING = auto()
+        """Current license is missing"""
+        CONNECTED = auto()
+        """Connected to server and ready to use API"""
+        RUNTIME_ERROR = auto()
+
+    def __init__(self, *, parent: QObject):
```

Comment:
What's the porpoise 🐬 of keeping `parent` from being a positional argument? Seems okay since it's the only argument.

Reply:
Well, QT will let you put most anything as the parent without any immediate errors, but it might lead to hard to figure out bugs later...

Reply:
It makes it a lot clearer what the random final argument in a constructor is. I've been standardizing on just making it a mandatory kwarg through the codebase for any QObject

Reply:
Okay!